internal/proxyservice/tcp/mysql/protocol/protocol.go

```diff
@@ -558,20 +559,31 @@ func UnpackHandshakeResponse41(packet []byte) (*HandshakeResponse41, error) {
 }
 
 // InjectCredentials takes in a HandshakeResponse41 from the client, the
-// salt from the server, and a username / password, and uses the salt
-// from the server handshake to inject the username / password credentials into
-// the client handshake response
-func InjectCredentials(clientHandshake *HandshakeResponse41, salt []byte, username string, password string) (err error) {
+// HandshakeV10 from the server, and a username / password, and uses the authData
+// from the server handshake to decide on and inject the username / password
+// credentials into the client handshake response
+func InjectCredentials(clientHandshake *HandshakeResponse41, serverHandshake *HandshakeV10, username string, password string) (err error) {
+	authData := serverHandshake.AuthData
+	authPluginName := serverHandshake.AuthPlugin
+
+	var authResponse []byte
+	switch authPluginName {
+	case "caching_sha2_password":
+		authResponse = ScrambleSHA256Password(authData, password)
+	case "mysql_native_password":
+		authResponse, err = NativePassword([]byte(password), authData)
+	default:
+		return errors.New("unsupported authentication protocol")
+	}
 
-	authResponse, err := NativePassword([]byte(password), salt)
```

Comment:
I know it's existing code but we should rewrite this to avoid naked returns 👇 :

Relevant style guide entry: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

Relevant quote from that link:

> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.